net: icmp: Add a way to send ICMP echo-req from system workqueue

jukkar opened the pull request:
Create a net_icmp_send_echo_request_from_work() variant that will not wait when allocating net_buf to be sent. This variant will avoid a warning to be printed when sending echo-req from a k_work handler because the function cannot sleep in this case.

jukkar committed:
net: icmp: Add a way to send ICMP Echo-Req without waiting

Create a net_icmp_send_echo_request_no_wait() variant that will
not wait when allocating net_buf to be sent. This variant will avoid
a warning to be printed when sending Echo-Req from a system workqueue
because the work cannot sleep in this case.

Signed-off-by: Jukka Rissanen <jukka.rissanen@nordicsemi.no>

rlubos commented on include/zephyr/net/icmp.h:
```diff
@@ -191,6 +191,30 @@ int net_icmp_send_echo_request(struct net_icmp_ctx *ctx,
+ * @brief Send ICMP echo request message from k_work handler.
+ *
+ * @details This function can be used to send ICMP echo request from
+ *          a k_work handler. This variant will not wait if the network
```
Technically this is only true for system workqueue, blocking on a custom workqueue should have no side effects (warnings printed).

rlubos commented on include/zephyr/net/icmp.h:
```diff
@@ -191,6 +191,30 @@ int net_icmp_send_echo_request(struct net_icmp_ctx *ctx,
+ *
+ * @return Return 0 if the sending succeed, <0 otherwise.
+ */
+int net_icmp_send_echo_request_from_work(struct net_icmp_ctx *ctx,
```
Personally I'm a bit puzzled with the function name, it describes a possible use-case for the function, and not the function behavior, I'd prefer something like `net_icmp_send_echo_request_no_wait()`. But no strong push though.

jukkar replied:
I was pondering with the name too, the `no_wait` variant gave a hint to me that there is no wait for echo-request command itself. But if it is clear what it indicates, we can certainly rename it.

jukkar replied:
After second thought, I will change this to `no_wait` which I had originally but then started to overthink the name too much.

rlubos reviewed (commented):
Shall this also fix the actual offender causing the warning to be printed (https://github.com/zephyrproject-rtos/zephyr/blob/798122e2ab5fc53402c91728e51221d4c91f16f1/subsys/net/lib/shell/ping.c#L275)?

jukkar commented:
> Shall this also fix the actual offender causing the warning to be printed (

No, that was missing as it was not clear to me what call is actually called from work handler.

jukkar committed:
net: shell: ping: Do not wait in ICMP Echo-Req

Use the net_icmp_send_echo_request_no_wait() when sending
ICMP Echo-Req so that we can avoid the warning message from
system workqueue handler.

Signed-off-by: Jukka Rissanen <jukka.rissanen@nordicsemi.no>

jukkar commented:
* updated according to comments

pdgendt commented on subsys/net/ip/icmp.c:
```diff
@@ -460,12 +467,40 @@ int net_icmp_send_echo_request(struct net_icmp_ctx *ctx,
+						  PKT_WAIT_TIME);
+}
+
+int net_icmp_send_echo_request_from_work(struct net_icmp_ctx *ctx,
```
Name does not match?

jukkar replied:
:facepalm:

jukkar commented:
Fixed the function name
